packages/core/react/src/WalletProvider.tsx

```diff
@@ -23,285 +24,158 @@ export interface WalletProviderProps {
+        [adaptersWithDefaultsInjected, walletName]
     );
-
-    // Setup and teardown event listeners when the adapter changes
     useEffect(() => {
-        if (adapter) {
-            adapter.on('connect', handleConnect);
-            adapter.on('disconnect', handleDisconnect);
-            adapter.on('error', handleError);
-            return () => {
-                adapter.off('connect', handleConnect);
-                adapter.off('disconnect', handleDisconnect);
-                adapter.off('error', handleError);
-            };
+        if (adapter == null) {
+            return;
         }
-    }, [adapter, handleConnect, handleDisconnect, handleError]);
-
-    // When the adapter changes, disconnect the old one
+        function handleDisconnect() {
+            if (isUnloading.current) {
+                return;
+            }
+            if (walletName === SolanaMobileWalletAdapterWalletName && getIsMobile(adapters)) {
+                // Leave the adapter selected in the event of a disconnection.
+                return;
+            }
+            setWalletName(null);
+        }
+        adapter.on('disconnect', handleDisconnect);
+        return () => {
+            adapter.off('disconnect', handleDisconnect);
+        };
+    }, [adapter, adapters, setWalletName, walletName]);
+    const handleAutoConnect = useMemo(() => {
+        if (autoConnect !== true || !adapter) {
+            return;
+        }
+        if (walletName === SolanaMobileWalletAdapterWalletName && getIsMobile(adapters)) {
+            return (adapter as SolanaMobileWalletAdapter).autoConnect_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.bind(adapter);
+        } else {
+            return adapter.connect.bind(adapter);
+        }
+    }, [adapter, adapters, autoConnect, walletName]);
     useEffect(() => {
+        if (adapter == null) {
+            return;
+        }
         return () => {
-            adapter?.disconnect();
+            if (
+                // Selecting a wallet other than the mobile wallet adapter is not
+                // sufficient reason to call `disconnect` on the mobile wallet adapter.
+                // Calling `disconnect` on the mobile wallet adapter causes the entire
+                // authorization store to be wiped.
+                adapter.name !== SolanaMobileWalletAdapterWalletName
+            ) {
+                adapter.disconnect();
+            }
         };
     }, [adapter]);
-
-    // If autoConnect is enabled, try to connect when the adapter changes and is ready
+    const isUnloading = useRef(false);
     useEffect(() => {
-        if (
-            isConnecting.current ||
-            connected ||
-            !autoConnect ||
-            !adapter ||
-            !(readyState === WalletReadyState.Installed || readyState === WalletReadyState.Loadable)
-        )
+        if (walletName === SolanaMobileWalletAdapterWalletName && getIsMobile(adapters)) {
```

Comment:
Is it worth memoizing `getIsMobile(adapters)` so effects etc. don't happen as much?

Reply:
Well, here's the thing. This is a bit clowny, because `getIsMobile` isn't actually a dependency of any of the effects. If, for instance, the user-agent changed during the life of the webpage (???) React would not re-render.

The only useful way to `useMemo()` `getIsMobile` is to have it fire some sort of event when the user-agent changes, which seems like mega YAGNI territory.

I'm making the assumption here that `getIsMobile` is a pure function of `adapters`.